### PR TITLE
8304759: Add BitMap iterators

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change which adds iteration over the set bit positions of a
BitMap via an iterator object.  This is in addition to the existing iterate()
and reverse_iterate() functions that apply a function to the set bit positions.

There are two iterator classes: BitMap::Iterator and BitMap::ReverseIterator.
Iteration uses a single, self-contained iterator object, rather than a pair of
iterator objects as is done with standard library iterators.  This is because
the desired form of iteration over bitmaps doesn't lend itself well to the
begin/end pair style.  Typical usage is something like

```
for (BitMap::Iterator it{map, beg, end}; !it.is_empty(); it.step()) {
  ... use it.index()
}
```

However, as a syntactic convenience, these iterator classes provide begin()
and end() member functions returning objects supporting the requirements for
range-based for loops.  By constraining these objects to this usage we can
cope with the "iterator pair" requirements.  Usage is something like

```
for (BitMap::idx_t index : BitMap::Iterator{map, beg, end}) {
  ...
}
```

Testing:
mach5 tier1, which includes running the updated gtests on all Oracle-supported
platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304759](https://bugs.openjdk.org/browse/JDK-8304759): Add BitMap iterators

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [379dfa82](https://git.openjdk.org/jdk/pull/13184/files/379dfa82b39a00298ed8f45005b45cba6da5b96e)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - Committer) ⚠️ Review applies to [379dfa82](https://git.openjdk.org/jdk/pull/13184/files/379dfa82b39a00298ed8f45005b45cba6da5b96e)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [379dfa82](https://git.openjdk.org/jdk/pull/13184/files/379dfa82b39a00298ed8f45005b45cba6da5b96e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13184/head:pull/13184` \
`$ git checkout pull/13184`

Update a local copy of the PR: \
`$ git checkout pull/13184` \
`$ git pull https://git.openjdk.org/jdk.git pull/13184/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13184`

View PR using the GUI difftool: \
`$ git pr show -t 13184`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13184.diff">https://git.openjdk.org/jdk/pull/13184.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/bitMap.hpp [hunk 1]
```diff
@@ -385,6 +394,155 @@ class BitMap {
...
+  // Returns true if the remaining iteration range is empty.
+  bool is_empty() const;
+
+  // Returns the index of the first set bit in the remaining iteration range.
+  // precondition: !is_empty()
+  idx_t first() const;
+
+  // Returns the index of the last set bit in the remaining iteration range.
+  // precondition: !is_empty()
+  idx_t last() const;
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.hpp:420]: The comments for `first()` and `last()` are a little misleading. The invariant that the `first()`/`last()` is a set bit is only true for `first()` in the forward iterator and for `last()` in the reverse iterator. 

It seems more like a contract that if you use `first()` then you must have created `IteratorImpl(const BitMap* map, idx_t beg, idx_t end)` with `beg` being a set bit. And if you use `last()` then you must have created it with `end` being a set bit.

**Review by xmas92 — APPROVED** [review]
lgtm. 

Just a comment about the comments for `first()`/`last()`

**Review by tschatzl — APPROVED** [review]

File: src/hotspot/share/utilities/bitMap.hpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/utilities/bitMap.hpp:420]: Yes, that's correct.  Some of the comments for IteratorImpl didn't get properly updated.  I'll clean that up.

**Comment by kimbarrett** [comment]
Thanks @xmas92 , @stefank , and @tschatzl for reviews.

**Pull request closed without merging**